File: back-end/src/main/java/nl/hva/backend/domain/models/game/board/location/locationImpl/Property.java
package nl.hva.backend.domain.models.game.board.location.locationImpl;

import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import nl.hva.backend.domain.models.game.board.GameBoard;
import nl.hva.backend.domain.models.game.board.location.OwnableLocation;

import javax.persistence.*;
import java.util.Objects;

@Entity
@NoArgsConstructor
@Getter
@Setter
public class Property extends OwnableLocation {
    private PropertyColor color;
    private String name;
    private double propertyDevelopmentCost;

    private double rentWithNoHouse;
    private double rentWithOneHouse;
    private double rentWithTwoHouses;
    private double rentWithThreeHouses;
    private double rentWithFourHouses;
    private double rentWithHotel;

    public Property(
            long id,
            int position,
            String description,
            GameBoard gameBoard,
            double initialPurchasePrice,
            PropertyColor color,
            String name,
            double propertyDevelopmentCost,
            double rentWithNoHouse,
            double rentWithOneHouse,
            double rentWithTwoHouses,
            double rentWithThreeHouses,
            double rentWithFourHouses,
            double rentWithHotel
    ) {
        super(id, position, description, gameBoard, initialPurchasePrice);
        this.color = color;
        this.name = name;
        this.propertyDevelopmentCost = propertyDevelopmentCost;
        this.rentWithNoHouse = rentWithNoHouse;
        this.rentWithOneHouse = rentWithOneHouse;
        this.rentWithTwoHouses = rentWithTwoHouses;
        this.rentWithThreeHouses = rentWithThreeHouses;
        this.rentWithFourHouses = rentWithFourHouses;
        this.rentWithHotel = rentWithHotel;
    }

    @Entity
    @Getter
    @NoArgsConstructor
    public static class PropertyDevelopmentStageTracker {
        @Id
        @GeneratedValue(strategy = GenerationType.IDENTITY)
        private long id;

        @ManyToOne
        private Property property;

        @Setter
        private PropertyDevelopmentStage developmentStage;

        public PropertyDevelopmentStageTracker(Property property) {
            this.property = property;
            this.developmentStage = PropertyDevelopmentStage.NO_DEVELOPMENT;
        }

        /**
         * Checks whether this property belongs to this Tracker.
         *
         * @param property The property to be assessed.
         * @return Whether this property belongs to this Tracker.
         */
        public boolean isOf(Property property) {
            return Objects.equals(this.getProperty(), property);
        }

    }

    @Getter
    public enum PropertyColor {
        LIGHT_BLUE,
        DARK_PURPLE,
        ORANGE,
        PURPLE,
        RED,
        YELLOW,
        GREEN,
        DARK_BLUE
    }

    public enum PropertyDevelopmentStage {
        NO_DEVELOPMENT(0),
        ONE_HOUSE(1),
        TWO_HOUSES(2),
        THREE_HOUSES(3),
        FOUR_HOUSES(4),
        HOTEL(5);

        public final int STAGE;

        PropertyDevelopmentStage(int STAGE) {
            this.STAGE = STAGE;
        }
    }

}

File: front-end/src/app/domain/models/game/board/location/locationImpl/Property.ts
import {OwnableLocation} from "../OwnableLocation";

export enum PropertyDevelopmentStage {
  NO_DEVELOPMENT = "NO_DEVELOPMENT",
  ONE_HOUSE = "ONE_HOUSE",
  TWO_HOUSES = "TWO_HOUSES",
  THREE_HOUSES = "THREE_HOUSES",
  FOUR_HOUSES = "FOUR_HOUSES",
  HOTEL = "HOTEL"
}

export enum PropertyColor {
  LIGHT_BLUE = "light-blue",
  DARK_PURPLE = "dark-purple",
  ORANGE = "orange",
  PURPLE = "purple",
  RED = "red",
  YELLOW = "yellow",
  GREEN = "green",
  DARK_BLUE = "dark-blue",
}

export class Property extends OwnableLocation {
  public static readonly DEVELOPMENT_LEVELS = [
    {level: 0, key: PropertyDevelopmentStage.NO_DEVELOPMENT, value: "no development"},
    {level: 1, key: PropertyDevelopmentStage.ONE_HOUSE, value: "1 house"},
    {level: 2, key: PropertyDevelopmentStage.TWO_HOUSES, value: "2 houses"},
    {level: 3, key: PropertyDevelopmentStage.THREE_HOUSES, value: "3 houses"},
    {level: 4, key: PropertyDevelopmentStage.FOUR_HOUSES, value: "4 houses"},
    {level: 5, key: PropertyDevelopmentStage.HOTEL, value: "1 Hotel (No houses)"}
  ];

  private _color: PropertyColor;
  private _name: string;
  private _propertyDevelopmentCost: number;
  private _rentWithNoHouse: number;
  private _rentWithOneHouse: number;
  private _rentWithTwoHouses: number;
  private _rentWithThreeHouses: number;
  private _rentWithFourHouses: number;
  private _rentWithHotel: number;

  public constructor(
    id: number,
    position: number,
    description: string,
    initialPurchasePrice: number,
    color: PropertyColor,
    name: string,
    mortgageValue: number,
    propertyDevelopmentCost: number,
    rentWithNoHouse: number,
    rentWithOneHouse: number,
    rentWithTwoHouses: number,
    rentWithThreeHouses: number,
    rentWithFourHouses: number,
    rentWithHotel: number
  ) {
    super(Property.name, id, position, description, initialPurchasePrice, mortgageValue);
    this._color = color;
    this._name = name;
    this._propertyDevelopmentCost = propertyDevelopmentCost;
    this._rentWithNoHouse = rentWithNoHouse;
    this._rentWithOneHouse = rentWithOneHouse;
    this._rentWithTwoHouses = rentWithTwoHouses;
    this._rentWithThreeHouses = rentWithThreeHouses;
    this._rentWithFourHouses = rentWithFourHouses;
    this._rentWithHotel = rentWithHotel;
  }

  public get color(): PropertyColor {
    return this._color;
  }

  public static trueCopy(location: Property): Property | null {
    return new Property(
      location.id,
      location.position,
      location.description,
      location.initialPurchasePrice,
      PropertyColor[location.color],
      location.name,
      location.mortgageValue,
      location.propertyDevelopmentCost,
      location.rentWithNoHouse,
      location.rentWithOneHouse,
      location.rentWithTwoHouses,
      location.rentWithThreeHouses,
      location.rentWithFourHouses,
      location.rentWithHotel,
    );
  }

  public override toJson(): object {
    return Object.assign(
      super.toJson(),
      {
        "color": this.color,
        "name": this.name,
        "mortgageValue": this.mortgageValue,
        "propertyDevelopmentCost": this.propertyDevelopmentCost,
        "rentWithNoHouse": this.rentWithNoHouse,
        "rentWithOneHouse": this.rentWithOneHouse,
        "rentWithTwoHouses": this.rentWithTwoHouses,
        "rentWithThreeHouses": this.rentWithThreeHouses,
        "rentWithFourHouses": this.rentWithFourHouses,
        "rentWithHotel": this.rentWithHotel,
      }
    );
  }

  public get name(): string {
    return this._name;
  }

  public get propertyDevelopmentCost(): number {
    return this._propertyDevelopmentCost;
  }

  public get rentWithNoHouse(): number {
    return this._rentWithNoHouse;
  }

  public get rentWithOneHouse(): number {
    return this._rentWithOneHouse;
  }

  public get rentWithTwoHouses(): number {
    return this._rentWithTwoHouses;
  }

  public get rentWithThreeHouses(): number {
    return this._rentWithThreeHouses;
  }

  public get rentWithFourHouses(): number {
    return this._rentWithFourHouses;
  }

  public get rentWithHotel(): number {
    return this._rentWithHotel;
  }

  public static propertyDevelopmentStageDescriber(stage: PropertyDevelopmentStage): string {
    switch (stage) {
      case PropertyDevelopmentStage.NO_DEVELOPMENT:
        return "No house or hotel"
      case PropertyDevelopmentStage.ONE_HOUSE:
        return "One house"
      case PropertyDevelopmentStage.TWO_HOUSES:
        return "Tho houses"
      case PropertyDevelopmentStage.THREE_HOUSES:
        return "Three houses"
      case PropertyDevelopmentStage.FOUR_HOUSES:
        return "Four houses"
      case PropertyDevelopmentStage.HOTEL:
        return "One hotel"
      default:
        return ""
    }
  }

  public override toString(): string {
    return this.name;
  }

}
